// @alyle/ui/toolbar/bundles/alyle-ui-toolbar.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@alyle/ui'), require('@angular/common')) :
	typeof define === 'function' && define.amd ? define('ly.toolbar', ['exports', '@angular/core', '@alyle/ui', '@angular/common'], factory) :
	(factory((global.ly = global.ly || {}, global.ly.toolbar = {}),global.ng.core,global.ly.core,global.ng.common));
}(this, (function (exports,core,ui,common) { 'use strict';

var ToolbarItem = /** @class */ (function () {
    function ToolbarItem() {
    }
    return ToolbarItem;
}());
ToolbarItem.decorators = [
    { type: core.Directive, args: [{
                selector: 'ly-toolbar-item',
            },] },
];
var LyToolbar = /** @class */ (function () {
    function LyToolbar(bgAndColor) {
        if (bgAndColor) {
            bgAndColor.setAutoContrast();
        }
    }
    return LyToolbar;
}());
LyToolbar.decorators = [
    { type: core.Component, args: [{
                selector: 'ly-toolbar',
                template: '<ng-content></ng-content>',
                styles: [":host[ly-dense]{height:48px;line-height:48px}:host{background:0 0;margin:0;max-width:100%;width:100%;min-height:48px;height:64px;-webkit-box-sizing:border-box;box-sizing:border-box;display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-align:center;-ms-flex-align:center;align-items:center;position:relative;font-size:20px}", ":host /deep/ ly-toolbar-item{font-weight:400;vertical-align:middle;display:inline-block;overflow:hidden;text-overflow:ellipsis;white-space:nowrap}:host /deep/ ly-toolbar-item[ly-toolbar-item-padding]{padding:0 16px}"],
                changeDetection: core.ChangeDetectionStrategy.OnPush
            },] },
];
LyToolbar.ctorParameters = function () { return [
    { type: ui.LyBgColorAndRaised, decorators: [{ type: core.Optional },] },
]; };
var LyToolbarModule = /** @class */ (function () {
    function LyToolbarModule() {
    }
    return LyToolbarModule;
}());
LyToolbarModule.decorators = [
    { type: core.NgModule, args: [{
                imports: [common.CommonModule],
                exports: [LyToolbar, ToolbarItem],
                declarations: [LyToolbar, ToolbarItem]
            },] },
];

exports.ToolbarItem = ToolbarItem;
exports.LyToolbar = LyToolbar;
exports.LyToolbarModule = LyToolbarModule;

Object.defineProperty(exports, '__esModule', { value: true });

})));
