// @alyle/ui/carousel/esm5/alyle-ui-carousel.js
import { Injectable, Component, ContentChildren, Input, ElementRef, ChangeDetectionStrategy, ChangeDetectorRef, HostBinding, Optional, forwardRef, Inject, PLATFORM_ID, NgModule, defineInjectable, inject } from '@angular/core';
import { LyTheme, ProvidedInTheme, LyCommonModule, AlyleUIModule } from '@alyle/ui';
import { isPlatformBrowser, CommonModule } from '@angular/common';
import { DomSanitizer } from '@angular/platform-browser';
import { LyRippleModule } from '@alyle/ui/ripple';

var CarouselService = /** @class */ (function () {
    function CarouselService(theme) {
        this.theme = theme;
        this.classes = {
            root: this.theme.setStyle('carousel', function () { return (""); })
        };
    }
    return CarouselService;
}());
CarouselService.decorators = [
    { type: Injectable, args: [ProvidedInTheme,] },
];
CarouselService.ctorParameters = function () { return [
    { type: LyTheme, },
]; };
CarouselService.ngInjectableDef = defineInjectable({ factory: function CarouselService_Factory() { return new CarouselService(inject(LyTheme)); }, token: CarouselService, providedIn: AlyleUIModule });
var LyCarousel = /** @class */ (function () {
    function LyCarousel(elementRef, sanitizer, cd, platformId) {
        this.elementRef = elementRef;
        this.sanitizer = sanitizer;
        this.cd = cd;
        this.platformId = platformId;
        this._interval = 60000;
        this.nullImg = 'data:image/gif;base64,R0lGODlhAQABAIABAP///wAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==';
        this._itemSelect = 0;
    }
    LyCarousel.prototype.sanitizerStyle = function (val) {
        return this.sanitizer.bypassSecurityTrustStyle(val);
    };
    Object.defineProperty(LyCarousel.prototype, "interval", {
        get: function () { return this._interval; },
        set: function (value) {
            this._interval = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LyCarousel.prototype, "stateIndicator", {
        get: function () {
            if (this.lyItems === undefined) {
                return 0;
            }
            else {
                var st = (this._itemSelect + 1) * 100 / this.lyItems.length;
                return st;
            }
        },
        enumerable: true,
        configurable: true
    });
    LyCarousel.prototype.ngOnDestroy = function () {
        if (isPlatformBrowser(this.platformId)) {
            clearInterval(this._fnInterval);
        }
    };
    LyCarousel.prototype._intervalCarousel = function (value) {
        var _this = this;
        if (value === void 0) { value = '+'; }
        if (isPlatformBrowser(this.platformId)) {
            if (value === '+') {
                this._itemSelect++;
                this._itemSelect = (this._itemSelect === (this.lyItems.length) ? 0 : this._itemSelect);
            }
            else if (value === '-') {
                this._itemSelect--;
                this._itemSelect = (this._itemSelect <= -1 ? (this.lyItems.length - 1) : this._itemSelect--);
            }
            else {
                this._itemSelect = value;
            }
            var intrval$ = {
                interval$: setInterval(function () {
                    _this._itemSelect++;
                    _this._itemSelect = (_this._itemSelect === (_this.lyItems.length) ? 0 : _this._itemSelect++);
                    _this.setActiveItem();
                }, this._interval)
            };
            clearInterval(this._fnInterval);
            this._fnInterval = intrval$.interval$;
            this.setActiveItem();
        }
    };
    LyCarousel.prototype.setActiveItem = function () {
        var _this = this;
        Promise.resolve(null).then(function () {
            var controlsBottom = _this.elementRef.nativeElement.querySelector('.carousel-indicators-container');
            _this.lyItems.forEach(function (_item) {
                _item.lyCarouselActive = false;
                _item._markForCheck();
            });
            var item = _this.lyItems.find(function (a, b) { return b === _this._itemSelect; });
            item.lyCarouselActive = true;
            _this.cd.markForCheck();
        });
    };
    LyCarousel.prototype._markForCheck = function () {
        this.cd.markForCheck();
    };
    LyCarousel.prototype.ngAfterViewInit = function () {
        var _this = this;
        if (isPlatformBrowser(this.platformId)) {
            this._intervalCarousel(0);
            this.lyItems.changes.subscribe(function (carousel) {
                _this._itemSelect = 0;
                _this._intervalCarousel(0);
                _this.setActiveItem();
                _this.cd.markForCheck();
            });
        }
    };
    LyCarousel.prototype.select = function (val) {
        this._intervalCarousel(val);
    };
    LyCarousel.prototype.focusRight = function () {
        this._intervalCarousel('+');
    };
    LyCarousel.prototype.focusLeft = function () {
        this._intervalCarousel('-');
    };
    return LyCarousel;
}());
LyCarousel.decorators = [
    { type: Component, args: [{
                selector: 'ly-carousel',
                styles: ["@-webkit-keyframes animationIntroStart{0%{opacity:0}100%{opacity:1}}@keyframes animationIntroStart{0%{opacity:0}100%{opacity:1}}.ly-carousel-container{display:inherit;position:absolute}:host{display:inline-block}:host /deep/{-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;position:relative}:host /deep/ .lycarousel-slide{position:relative;overflow:hidden;width:inherit;height:inherit}:host /deep/ .lycarousel-slide .lycarousel-items{height:inherit;width:inherit}:host /deep/ .lycarousel-slide .lycarousel-items .lycarousel-item img{display:block;max-width:100%;height:inherit;width:inherit;background-repeat:no-repeat;background-size:cover;background-position:center;position:absolute}:host /deep/ .lycarousel-item{-webkit-animation:450ms animationIntroStart;animation:450ms animationIntroStart;height:inherit;width:inherit;-webkit-transition:-webkit-transform .6s cubic-bezier(.65,.05,.36,1);transition:-webkit-transform .6s cubic-bezier(.65,.05,.36,1);transition:transform .6s cubic-bezier(.65,.05,.36,1);transition:transform .6s cubic-bezier(.65,.05,.36,1),-webkit-transform .6s cubic-bezier(.65,.05,.36,1);-webkit-backface-visibility:hidden;backface-visibility:hidden;-webkit-perspective:1000px;perspective:1000px;display:none;position:relative;-webkit-transition:.6s cubic-bezier(.65,.05,.36,1) left;transition:.6s cubic-bezier(.65,.05,.36,1) left}:host /deep/ .lycarousel-item.active{display:block;left:0}:host /deep/ .lycarousel-content{position:absolute;top:0;bottom:0;left:0;right:0;width:100%;height:100%}:host /deep/ .lycarousel-cursor-left,:host /deep/ .lycarousel-cursor-right{position:absolute;top:0;bottom:0;width:50%}:host /deep/ svg{height:0;width:0;position:absolute}:host /deep/ .carousel-indicators-container{z-index:1;height:48px;width:100%;overflow:hidden;position:absolute;bottom:0}:host /deep/ .carousel-blur{-webkit-animation:725ms animationIntroStart;animation:725ms animationIntroStart;height:220px;width:100%;overflow:hidden;-webkit-transform:translate(0,0);transform:translate(0,0);position:absolute;background-size:cover;background-position:center;bottom:0;filter:blur(15px);-webkit-filter:blur(15px);-moz-filter:blur(15px);-o-filter:blur(15px);-ms-filter:blur(15px);opacity:.75}:host /deep/ .carousel-indicators{padding:0;z-index:1;position:absolute;bottom:0;width:100%;text-align:center;margin:0;-webkit-box-sizing:border-box;box-sizing:border-box;background:inherit;display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-align:center;-ms-flex-align:center;align-items:center;-webkit-box-pack:center;-ms-flex-pack:center;justify-content:center;height:48px}:host /deep/ .carousel-indicators li{display:inline-block;border-radius:50%;cursor:pointer;position:relative;padding:.5em}:host /deep/ .carousel-indicators li span{-webkit-transition:.3s cubic-bezier(.65,.05,.36,1);transition:.3s cubic-bezier(.65,.05,.36,1);width:1em;height:1em;-webkit-transform:scale(.5);transform:scale(.5);border-radius:50%;display:block}:host /deep/ .carousel-indicators li span.active{-webkit-transform:scale(1);transform:scale(1)}:host /deep/ .carousel-indicators li:focus{outline:0}:host /deep/ .carousel-indicators li:active{outline:0}.ly-carousel-inactive-button{background:#fff!important}"],
                template: "\n  <div class=\"lycarousel-slide\">\n    <!--buttons-->\n    <div class=\"carousel-indicators-container\">\n      <div class=\"carousel-blur\" [style.background-image]=\"'url('+lyItems.toArray()[_itemSelect]?.src+')'\"></div>\n    </div>\n    <ul class=\"carousel-indicators\" *ngIf=\"lyItems.length!=1\">\n      <li tabindex=\"0\"\n      (click)=\"select(i)\"\n      role=\"button\"\n      *ngFor=\"let item of lyItems; let i = index\">\n      <span color=\"#000\"\n      bg=\"#fff\"\n      [class.active]=\"_itemSelect==i\"\n      [newRaised]=\"[6]\"></span>\n      </li>\n    </ul>\n    <!--items-->\n    <div class=\"lycarousel-items\">\n      <div [class.active]=\"_itemSelect==i\"\n      class=\"lycarousel-item\"\n      *ngFor=\"let item of lyItems; let i = index\">\n        <img [src]=\"nullImg\"\n        [style.background-image]=\"'url('+item.src+')'\">\n      </div>\n      <!--<div class=\"lycarousel-content\">\n      <ng-content></ng-content>\n      </div>-->\n    </div>\n    <!--cursor-->\n  </div>\n  <div class=\"lycarousel-content ly-carousel-container\">\n    <ng-content></ng-content>\n  </div>\n  ",
                changeDetection: ChangeDetectionStrategy.OnPush,
                preserveWhitespaces: false
            },] },
];
LyCarousel.ctorParameters = function () { return [
    { type: ElementRef, },
    { type: DomSanitizer, },
    { type: ChangeDetectorRef, },
    { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] },] },
]; };
LyCarousel.propDecorators = {
    "lyItems": [{ type: ContentChildren, args: [forwardRef(function () { return LyCarouselItemComponent; }),] },],
    "interval": [{ type: Input, args: ['interval',] },],
};
var LyCarouselItemComponent = /** @class */ (function () {
    function LyCarouselItemComponent(carousel, carouselService, cd, platformId) {
        this.carouselService = carouselService;
        this.cd = cd;
        this.platformId = platformId;
        this.lyCarouselActive = false;
        this._carousel = carousel;
    }
    LyCarouselItemComponent.prototype.ngOnChanges = function (changes) { };
    LyCarouselItemComponent.prototype.ngOnInit = function () { };
    LyCarouselItemComponent.prototype._markForCheck = function () {
        this.cd.markForCheck();
    };
    return LyCarouselItemComponent;
}());
LyCarouselItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'ly-carousel-item',
                styles: ["\n    :host:not(.ly-carousel-item-active) {\n      display: none !important;\n    }\n  "],
                template: "\n  <ng-template [ngIf]=\"lyCarouselActive\">\n    <ng-content></ng-content>\n  </ng-template>\n  ",
                changeDetection: ChangeDetectionStrategy.OnPush
            },] },
];
LyCarouselItemComponent.ctorParameters = function () { return [
    { type: LyCarousel, decorators: [{ type: Optional },] },
    { type: CarouselService, },
    { type: ChangeDetectorRef, },
    { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] },] },
]; };
LyCarouselItemComponent.propDecorators = {
    "src": [{ type: Input },],
    "lyCarouselActive": [{ type: Input }, { type: HostBinding, args: ['class.ly-carousel-item-active',] },],
};
var LY_CAROUSEL_DIRECTIVES = [LyCarouselItemComponent, LyCarousel];
var LyCarouselModule = /** @class */ (function () {
    function LyCarouselModule() {
    }
    return LyCarouselModule;
}());
LyCarouselModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, LyCommonModule, LyRippleModule],
                exports: [LY_CAROUSEL_DIRECTIVES],
                declarations: [LY_CAROUSEL_DIRECTIVES]
            },] },
];

export { LyCarousel, LyCarouselItemComponent, CarouselService, LyCarouselModule };
