// @alyle/ui/checkbox/bundles/alyle-ui-checkbox.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@angular/common')) :
	typeof define === 'function' && define.amd ? define('ly.checkbox', ['exports', '@angular/core', '@angular/forms', '@angular/common'], factory) :
	(factory((global.ly = global.ly || {}, global.ly.checkbox = {}),global.ng.core,global.ng.forms,global.ng.common));
}(this, (function (exports,core,forms,common) { 'use strict';

var LY_CHECKBOX_CONTROL_VALUE_ACCESSOR = {
    provide: forms.NG_VALUE_ACCESSOR,
    useExisting: core.forwardRef(function () { return LyCheckbox; }),
    multi: true
};
var CheckboxState = {
    Init: 0,
    Checked: 1,
    Unchecked: 2,
    Indeterminate: 3,
};
CheckboxState[CheckboxState.Init] = "Init";
CheckboxState[CheckboxState.Checked] = "Checked";
CheckboxState[CheckboxState.Unchecked] = "Unchecked";
CheckboxState[CheckboxState.Indeterminate] = "Indeterminate";
var LyCheckbox = /** @class */ (function () {
    function LyCheckbox() {
        this._currentCheckState = CheckboxState.Init;
        this.checked = false;
        this.change = new core.EventEmitter();
        this.indeterminateChange = new core.EventEmitter();
        this.onTouched = function () { };
        this._controlValueAccessorChangeFn = function (value) { };
    }
    LyCheckbox.prototype.toggle = function () {
        this.checked = !this.checked;
    };
    LyCheckbox.prototype._onChangeEvent = function (event) {
        event.stopPropagation();
    };
    LyCheckbox.prototype._onInputClick = function (event) {
        var ev = (event.target);
        this.checked = ev.checked;
        this._controlValueAccessorChangeFn(this.checked);
        console.log('_onInputClick', ev.checked);
    };
    LyCheckbox.prototype.writeValue = function (value) {
        this.checked = !!value;
        console.log('value', this.checked);
    };
    LyCheckbox.prototype.registerOnChange = function (fn) {
        this._controlValueAccessorChangeFn = fn;
    };
    LyCheckbox.prototype.ngOnChanges = function (changes) {
        console.log('changes', changes);
        if (changes['src']) {
        }
    };
    LyCheckbox.prototype.registerOnTouched = function (fn) {
        this.onTouched = fn;
    };
    LyCheckbox.prototype.ngOnInit = function () {
    };
    LyCheckbox.prototype.ngAfterContentInit = function () {
    };
    LyCheckbox.prototype.ngOnDestroy = function () {
    };
    return LyCheckbox;
}());
LyCheckbox.decorators = [
    { type: core.Component, args: [{
                selector: 'ly-checkbox',
                template: "<label>\n  <div class=\"ly-checkbox-container\">\n    <input #input\n           [hidden]=\"false\" type=\"checkbox\"\n           [id]=\"id\"\n           [required]=\"required\"\n           [checked]=\"checked\"\n           [value]=\"value\"\n           [disabled]=\"disabled\"\n           [name]=\"name\"\n           [indeterminate]=\"indeterminate\"\n           [attr.aria-label]=\"ariaLabel\"\n           [attr.aria-labelledby]=\"ariaLabelledby\"\n           (change)=\"_onChangeEvent($event)\"\n           (click)=\"_onInputClick($event)\">\n  </div>\n</label>\n",
                styles: [":host{display:inline-block}"],
                providers: [LY_CHECKBOX_CONTROL_VALUE_ACCESSOR]
            },] },
];
LyCheckbox.ctorParameters = function () { return []; };
LyCheckbox.propDecorators = {
    "id": [{ type: core.Input, args: ['id',] },],
    "ariaLabel": [{ type: core.Input, args: ['attr.aria-label',] },],
    "ariaLabelledby": [{ type: core.Input, args: ['attr.aria-labelledby',] },],
    "required": [{ type: core.Input, args: ['required',] },],
    "labelPosition": [{ type: core.Input, args: ['labelPosition',] },],
    "color": [{ type: core.Input, args: ['color',] },],
    "value": [{ type: core.Input, args: ['value',] },],
    "name": [{ type: core.Input, args: ['name',] },],
    "disabled": [{ type: core.Input, args: ['disabled',] },],
    "checked": [{ type: core.Input, args: ['checked',] },],
    "indeterminate": [{ type: core.Input, args: ['indeterminate',] },],
    "change": [{ type: core.Output, args: ['change',] },],
    "indeterminateChange": [{ type: core.Output, args: ['indeterminateChange',] },],
};
var LyCheckboxService = /** @class */ (function () {
    function LyCheckboxService() {
    }
    return LyCheckboxService;
}());
LyCheckboxService.decorators = [
    { type: core.Injectable },
];
LyCheckboxService.ctorParameters = function () { return []; };
var LyCheckboxModule = /** @class */ (function () {
    function LyCheckboxModule() {
    }
    LyCheckboxModule.forRoot = function () {
        return {
            ngModule: LyCheckboxModule,
        };
    };
    return LyCheckboxModule;
}());
LyCheckboxModule.decorators = [
    { type: core.NgModule, args: [{
                imports: [
                    common.CommonModule
                ],
                exports: [LyCheckbox],
                declarations: [LyCheckbox],
                providers: [LyCheckboxService]
            },] },
];

exports.LyCheckboxModule = LyCheckboxModule;
exports.LyCheckboxService = LyCheckboxService;
exports.LY_CHECKBOX_CONTROL_VALUE_ACCESSOR = LY_CHECKBOX_CONTROL_VALUE_ACCESSOR;
exports.CheckboxState = CheckboxState;
exports.LyCheckbox = LyCheckbox;

Object.defineProperty(exports, '__esModule', { value: true });

})));
