// @alyle/ui/flex/esm5/alyle-ui-flex.js
import { Injectable, Directive, Input, Renderer2, ElementRef, NgModule, defineInjectable, inject } from '@angular/core';
import { LyTheme } from '@alyle/ui';

var LyFlexService = /** @class */ (function () {
    function LyFlexService(theme) {
        this.theme = theme;
        this.fxStyle = "box-sizing:border-box;display:flex;";
    }
    return LyFlexService;
}());
LyFlexService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] },
];
LyFlexService.ctorParameters = function () { return [
    { type: LyTheme, },
]; };
LyFlexService.ngInjectableDef = defineInjectable({ factory: function LyFlexService_Factory() { return new LyFlexService(inject(LyTheme)); }, token: LyFlexService, providedIn: "root" });
var LyFlex = /** @class */ (function () {
    function LyFlex(renderer, elementRef, fxService) {
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.fxService = fxService;
        this._display = 'flex';
    }
    Object.defineProperty(LyFlex.prototype, "fxInline", {
        set: function (val) {
            this._display = 'inline-flex';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LyFlex.prototype, "fxAlign", {
        set: function (val) {
            var justifyContent = val[0] || 'flex-start';
            var alignItems = val[1] || 'stretch';
            var alignContent = val[2];
            this._fxAlign = ("justify-content:" + justifyContent + ";" +
                ("align-items:" + alignItems + ";") +
                ("align-content:" + (alignContent || alignItems) + ";"));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LyFlex.prototype, "fxDirection", {
        set: function (val) {
            this._fxDirection = "flex-direction:" + (val || 'row') + ";";
        },
        enumerable: true,
        configurable: true
    });
    LyFlex.prototype.ngOnChanges = function (changes) {
        var _this = this;
        var key = '';
        var styles = "";
        for (var inputKey in changes) {
            key += inputKey + changes[inputKey].currentValue;
            styles += this["_" + inputKey] || '';
        }
        var classname = this.fxService.theme.setStyle(key, function () { return ("display:" + _this._display + ";" +
            styles); });
        this.fxService.theme.updateClassName(this.elementRef.nativeElement, this.renderer, classname, this._currentClassname);
    };
    return LyFlex;
}());
LyFlex.decorators = [
    { type: Directive, args: [{
                selector: '[fxDirection], [fxWrap], [fxAlign], [fxInline]'
            },] },
];
LyFlex.ctorParameters = function () { return [
    { type: Renderer2, },
    { type: ElementRef, },
    { type: LyFlexService, },
]; };
LyFlex.propDecorators = {
    "fxInline": [{ type: Input },],
    "fxAlign": [{ type: Input },],
    "fxDirection": [{ type: Input },],
    "fxWrap": [{ type: Input },],
};
var LyFlexModule = /** @class */ (function () {
    function LyFlexModule() {
    }
    return LyFlexModule;
}());
LyFlexModule.decorators = [
    { type: NgModule, args: [{
                imports: [],
                exports: [LyFlex],
                declarations: [LyFlex],
                providers: [],
            },] },
];

export { LyFlex, LyFlexModule, LyFlexService as ɵa };
