// @alyle/ui/carousel/esm2015/alyle-ui-carousel.js
import { Injectable, Component, ContentChildren, Input, ElementRef, ChangeDetectionStrategy, ChangeDetectorRef, HostBinding, Optional, forwardRef, Inject, PLATFORM_ID, NgModule, defineInjectable, inject } from '@angular/core';
import { LyTheme, ProvidedInTheme, LyCommonModule, AlyleUIModule } from '@alyle/ui';
import { isPlatformBrowser, CommonModule } from '@angular/common';
import { DomSanitizer } from '@angular/platform-browser';
import { LyRippleModule } from '@alyle/ui/ripple';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class CarouselService {
    /**
     * @param {?} theme
     */
    constructor(theme) {
        this.theme = theme;
        this.classes = {
            root: this.theme.setStyle('carousel', () => (``))
        };
    }
}
CarouselService.decorators = [
    { type: Injectable, args: [ProvidedInTheme,] },
];
/** @nocollapse */
CarouselService.ctorParameters = () => [
    { type: LyTheme, },
];
/** @nocollapse */ CarouselService.ngInjectableDef = defineInjectable({ factory: function CarouselService_Factory() { return new CarouselService(inject(LyTheme)); }, token: CarouselService, providedIn: AlyleUIModule });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class LyCarousel {
    /**
     * @param {?} elementRef
     * @param {?} sanitizer
     * @param {?} cd
     * @param {?} platformId
     */
    constructor(elementRef, sanitizer, cd, platformId) {
        this.elementRef = elementRef;
        this.sanitizer = sanitizer;
        this.cd = cd;
        this.platformId = platformId;
        this._interval = 60000;
        this.nullImg = 'data:image/gif;base64,R0lGODlhAQABAIABAP///wAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==';
        this._itemSelect = 0;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    sanitizerStyle(val) {
        return this.sanitizer.bypassSecurityTrustStyle(val);
    }
    /**
     * @return {?}
     */
    get interval() { return this._interval; }
    /**
     * @param {?} value
     * @return {?}
     */
    set interval(value) {
        this._interval = value;
    }
    /**
     * @return {?}
     */
    get stateIndicator() {
        if (this.lyItems === undefined) {
            return 0;
        }
        else {
            const /** @type {?} */ st = (this._itemSelect + 1) * 100 / this.lyItems.length;
            return st;
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (isPlatformBrowser(this.platformId)) {
            clearInterval(this._fnInterval);
        }
    }
    /**
     * @param {?=} value
     * @return {?}
     */
    _intervalCarousel(value = '+') {
        if (isPlatformBrowser(this.platformId)) {
            if (value === '+') {
                this._itemSelect++;
                this._itemSelect = (this._itemSelect === (this.lyItems.length) ? 0 : this._itemSelect);
                // console.log(this._itemSelect, this.lyItems.length);
            }
            else if (value === '-') {
                this._itemSelect--;
                this._itemSelect = (this._itemSelect <= -1 ? (this.lyItems.length - 1) : this._itemSelect--);
                // console.log('--',this._itemSelect, this.lyItems.length);
            }
            else {
                this._itemSelect = value;
            }
            const /** @type {?} */ intrval$ = {
                interval$: setInterval(() => {
                    this._itemSelect++;
                    this._itemSelect = (this._itemSelect === (this.lyItems.length) ? 0 : this._itemSelect++);
                    // console.log('interval state', this._itemSelect);
                    this.setActiveItem();
                }, this._interval)
            };
            clearInterval(this._fnInterval);
            this._fnInterval = intrval$.interval$;
            this.setActiveItem();
        }
    }
    /**
     * @return {?}
     */
    setActiveItem() {
        // let activeItems = this.lyItems.filter((tab)=>tab.active);
        Promise.resolve(null).then(() => {
            const /** @type {?} */ controlsBottom = this.elementRef.nativeElement.querySelector('.carousel-indicators-container');
            // controlsBottom.classList.remove('animation');
            // // tslint:disable-next-line:no-unused-expression
            // void controlsBottom.offsetWidth;
            // controlsBottom.classList.add('animation');
            this.lyItems.forEach((_item) => {
                _item.lyCarouselActive = false;
                _item._markForCheck();
            });
            const /** @type {?} */ item = this.lyItems.find((a, b) => b === this._itemSelect);
            item.lyCarouselActive = true;
            this.cd.markForCheck();
        });
    }
    /**
     * @return {?}
     */
    _markForCheck() {
        this.cd.markForCheck();
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (isPlatformBrowser(this.platformId)) {
            this._intervalCarousel(0);
            this.lyItems.changes.subscribe((carousel) => {
                this._itemSelect = 0;
                this._intervalCarousel(0);
                this.setActiveItem();
                this.cd.markForCheck();
            });
        }
    }
    /**
     * @param {?} val
     * @return {?}
     */
    select(val) {
        this._intervalCarousel(val);
    }
    /**
     * @return {?}
     */
    focusRight() {
        this._intervalCarousel('+');
    }
    /**
     * @return {?}
     */
    focusLeft() {
        this._intervalCarousel('-');
    }
}
LyCarousel.decorators = [
    { type: Component, args: [{
                selector: 'ly-carousel',
                styles: [`@-webkit-keyframes animationIntroStart{0%{opacity:0}100%{opacity:1}}@keyframes animationIntroStart{0%{opacity:0}100%{opacity:1}}.ly-carousel-container{display:inherit;position:absolute}:host{display:inline-block}:host /deep/{-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;position:relative}:host /deep/ .lycarousel-slide{position:relative;overflow:hidden;width:inherit;height:inherit}:host /deep/ .lycarousel-slide .lycarousel-items{height:inherit;width:inherit}:host /deep/ .lycarousel-slide .lycarousel-items .lycarousel-item img{display:block;max-width:100%;height:inherit;width:inherit;background-repeat:no-repeat;background-size:cover;background-position:center;position:absolute}:host /deep/ .lycarousel-item{-webkit-animation:450ms animationIntroStart;animation:450ms animationIntroStart;height:inherit;width:inherit;-webkit-transition:-webkit-transform .6s cubic-bezier(.65,.05,.36,1);transition:-webkit-transform .6s cubic-bezier(.65,.05,.36,1);transition:transform .6s cubic-bezier(.65,.05,.36,1);transition:transform .6s cubic-bezier(.65,.05,.36,1),-webkit-transform .6s cubic-bezier(.65,.05,.36,1);-webkit-backface-visibility:hidden;backface-visibility:hidden;-webkit-perspective:1000px;perspective:1000px;display:none;position:relative;-webkit-transition:.6s cubic-bezier(.65,.05,.36,1) left;transition:.6s cubic-bezier(.65,.05,.36,1) left}:host /deep/ .lycarousel-item.active{display:block;left:0}:host /deep/ .lycarousel-content{position:absolute;top:0;bottom:0;left:0;right:0;width:100%;height:100%}:host /deep/ .lycarousel-cursor-left,:host /deep/ .lycarousel-cursor-right{position:absolute;top:0;bottom:0;width:50%}:host /deep/ svg{height:0;width:0;position:absolute}:host /deep/ .carousel-indicators-container{z-index:1;height:48px;width:100%;overflow:hidden;position:absolute;bottom:0}:host /deep/ .carousel-blur{-webkit-animation:725ms animationIntroStart;animation:725ms animationIntroStart;height:220px;width:100%;overflow:hidden;-webkit-transform:translate(0,0);transform:translate(0,0);position:absolute;background-size:cover;background-position:center;bottom:0;filter:blur(15px);-webkit-filter:blur(15px);-moz-filter:blur(15px);-o-filter:blur(15px);-ms-filter:blur(15px);opacity:.75}:host /deep/ .carousel-indicators{padding:0;z-index:1;position:absolute;bottom:0;width:100%;text-align:center;margin:0;-webkit-box-sizing:border-box;box-sizing:border-box;background:inherit;display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-align:center;-ms-flex-align:center;align-items:center;-webkit-box-pack:center;-ms-flex-pack:center;justify-content:center;height:48px}:host /deep/ .carousel-indicators li{display:inline-block;border-radius:50%;cursor:pointer;position:relative;padding:.5em}:host /deep/ .carousel-indicators li span{-webkit-transition:.3s cubic-bezier(.65,.05,.36,1);transition:.3s cubic-bezier(.65,.05,.36,1);width:1em;height:1em;-webkit-transform:scale(.5);transform:scale(.5);border-radius:50%;display:block}:host /deep/ .carousel-indicators li span.active{-webkit-transform:scale(1);transform:scale(1)}:host /deep/ .carousel-indicators li:focus{outline:0}:host /deep/ .carousel-indicators li:active{outline:0}.ly-carousel-inactive-button{background:#fff!important}`],
                template: `
  <div class="lycarousel-slide">
    <!--buttons-->
    <div class="carousel-indicators-container">
      <div class="carousel-blur" [style.background-image]="'url('+lyItems.toArray()[_itemSelect]?.src+')'"></div>
    </div>
    <ul class="carousel-indicators" *ngIf="lyItems.length!=1">
      <li tabindex="0"
      (click)="select(i)"
      role="button"
      *ngFor="let item of lyItems; let i = index">
      <span color="#000"
      bg="#fff"
      [class.active]="_itemSelect==i"
      [newRaised]="[6]"></span>
      </li>
    </ul>
    <!--items-->
    <div class="lycarousel-items">
      <div [class.active]="_itemSelect==i"
      class="lycarousel-item"
      *ngFor="let item of lyItems; let i = index">
        <img [src]="nullImg"
        [style.background-image]="'url('+item.src+')'">
      </div>
      <!--<div class="lycarousel-content">
      <ng-content></ng-content>
      </div>-->
    </div>
    <!--cursor-->
  </div>
  <div class="lycarousel-content ly-carousel-container">
    <ng-content></ng-content>
  </div>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush,
                preserveWhitespaces: false
            },] },
];
/** @nocollapse */
LyCarousel.ctorParameters = () => [
    { type: ElementRef, },
    { type: DomSanitizer, },
    { type: ChangeDetectorRef, },
    { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] },] },
];
LyCarousel.propDecorators = {
    "lyItems": [{ type: ContentChildren, args: [forwardRef(() => LyCarouselItemComponent),] },],
    "interval": [{ type: Input, args: ['interval',] },],
};
class LyCarouselItemComponent {
    /**
     * @param {?} carousel
     * @param {?} carouselService
     * @param {?} cd
     * @param {?} platformId
     */
    constructor(carousel, carouselService, cd, platformId) {
        this.carouselService = carouselService;
        this.cd = cd;
        this.platformId = platformId;
        this.lyCarouselActive = false;
        this._carousel = carousel;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) { }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @return {?}
     */
    _markForCheck() {
        this.cd.markForCheck();
    }
}
LyCarouselItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'ly-carousel-item',
                styles: [`
    :host:not(.ly-carousel-item-active) {
      display: none !important;
    }
  `],
                template: `
  <ng-template [ngIf]="lyCarouselActive">
    <ng-content></ng-content>
  </ng-template>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush
            },] },
];
/** @nocollapse */
LyCarouselItemComponent.ctorParameters = () => [
    { type: LyCarousel, decorators: [{ type: Optional },] },
    { type: CarouselService, },
    { type: ChangeDetectorRef, },
    { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] },] },
];
LyCarouselItemComponent.propDecorators = {
    "src": [{ type: Input },],
    "lyCarouselActive": [{ type: Input }, { type: HostBinding, args: ['class.ly-carousel-item-active',] },],
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
const LY_CAROUSEL_DIRECTIVES = [LyCarouselItemComponent, LyCarousel];
class LyCarouselModule {
}
LyCarouselModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, LyCommonModule, LyRippleModule],
                exports: [LY_CAROUSEL_DIRECTIVES],
                declarations: [LY_CAROUSEL_DIRECTIVES]
            },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */

export { LyCarousel, LyCarouselItemComponent, CarouselService, LyCarouselModule };
