// @alyle/ui/toolbar/esm5/alyle-ui-toolbar.js
import { Component, Directive, Optional, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { LyBgColorAndRaised } from '@alyle/ui';
import { CommonModule } from '@angular/common';

var ToolbarItem = /** @class */ (function () {
    function ToolbarItem() {
    }
    return ToolbarItem;
}());
ToolbarItem.decorators = [
    { type: Directive, args: [{
                selector: 'ly-toolbar-item',
            },] },
];
var LyToolbar = /** @class */ (function () {
    function LyToolbar(bgAndColor) {
        if (bgAndColor) {
            bgAndColor.setAutoContrast();
        }
    }
    return LyToolbar;
}());
LyToolbar.decorators = [
    { type: Component, args: [{
                selector: 'ly-toolbar',
                template: '<ng-content></ng-content>',
                styles: [":host[ly-dense]{height:48px;line-height:48px}:host{background:0 0;margin:0;max-width:100%;width:100%;min-height:48px;height:64px;-webkit-box-sizing:border-box;box-sizing:border-box;display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-align:center;-ms-flex-align:center;align-items:center;position:relative;font-size:20px}", ":host /deep/ ly-toolbar-item{font-weight:400;vertical-align:middle;display:inline-block;overflow:hidden;text-overflow:ellipsis;white-space:nowrap}:host /deep/ ly-toolbar-item[ly-toolbar-item-padding]{padding:0 16px}"],
                changeDetection: ChangeDetectionStrategy.OnPush
            },] },
];
LyToolbar.ctorParameters = function () { return [
    { type: LyBgColorAndRaised, decorators: [{ type: Optional },] },
]; };
var LyToolbarModule = /** @class */ (function () {
    function LyToolbarModule() {
    }
    return LyToolbarModule;
}());
LyToolbarModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule],
                exports: [LyToolbar, ToolbarItem],
                declarations: [LyToolbar, ToolbarItem]
            },] },
];

export { ToolbarItem, LyToolbar, LyToolbarModule };
