// @alyle/ui/flex/bundles/alyle-ui-flex.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@alyle/ui')) :
	typeof define === 'function' && define.amd ? define('ly.flex', ['exports', '@angular/core', '@alyle/ui'], factory) :
	(factory((global.ly = global.ly || {}, global.ly.flex = {}),global.ng.core,global.ly.core));
}(this, (function (exports,core,ui) { 'use strict';

var LyFlexService = /** @class */ (function () {
    function LyFlexService(theme) {
        this.theme = theme;
        this.fxStyle = "box-sizing:border-box;display:flex;";
    }
    return LyFlexService;
}());
LyFlexService.decorators = [
    { type: core.Injectable, args: [{
                providedIn: 'root'
            },] },
];
LyFlexService.ctorParameters = function () { return [
    { type: ui.LyTheme, },
]; };
LyFlexService.ngInjectableDef = core.defineInjectable({ factory: function LyFlexService_Factory() { return new LyFlexService(core.inject(ui.LyTheme)); }, token: LyFlexService, providedIn: "root" });
var LyFlex = /** @class */ (function () {
    function LyFlex(renderer, elementRef, fxService) {
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.fxService = fxService;
        this._display = 'flex';
    }
    Object.defineProperty(LyFlex.prototype, "fxInline", {
        set: function (val) {
            this._display = 'inline-flex';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LyFlex.prototype, "fxAlign", {
        set: function (val) {
            var justifyContent = val[0] || 'flex-start';
            var alignItems = val[1] || 'stretch';
            var alignContent = val[2];
            this._fxAlign = ("justify-content:" + justifyContent + ";" +
                ("align-items:" + alignItems + ";") +
                ("align-content:" + (alignContent || alignItems) + ";"));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LyFlex.prototype, "fxDirection", {
        set: function (val) {
            this._fxDirection = "flex-direction:" + (val || 'row') + ";";
        },
        enumerable: true,
        configurable: true
    });
    LyFlex.prototype.ngOnChanges = function (changes) {
        var _this = this;
        var key = '';
        var styles = "";
        for (var inputKey in changes) {
            key += inputKey + changes[inputKey].currentValue;
            styles += this["_" + inputKey] || '';
        }
        var classname = this.fxService.theme.setStyle(key, function () { return ("display:" + _this._display + ";" +
            styles); });
        this.fxService.theme.updateClassName(this.elementRef.nativeElement, this.renderer, classname, this._currentClassname);
    };
    return LyFlex;
}());
LyFlex.decorators = [
    { type: core.Directive, args: [{
                selector: '[fxDirection], [fxWrap], [fxAlign], [fxInline]'
            },] },
];
LyFlex.ctorParameters = function () { return [
    { type: core.Renderer2, },
    { type: core.ElementRef, },
    { type: LyFlexService, },
]; };
LyFlex.propDecorators = {
    "fxInline": [{ type: core.Input },],
    "fxAlign": [{ type: core.Input },],
    "fxDirection": [{ type: core.Input },],
    "fxWrap": [{ type: core.Input },],
};
var LyFlexModule = /** @class */ (function () {
    function LyFlexModule() {
    }
    return LyFlexModule;
}());
LyFlexModule.decorators = [
    { type: core.NgModule, args: [{
                imports: [],
                exports: [LyFlex],
                declarations: [LyFlex],
                providers: [],
            },] },
];

exports.LyFlex = LyFlex;
exports.LyFlexModule = LyFlexModule;
exports.ɵa = LyFlexService;

Object.defineProperty(exports, '__esModule', { value: true });

})));
